crates/hir_def/src/nameres/collector.rs

```diff
@@ -1468,16 +1469,28 @@ impl ModCollector<'_, '_> {
                     )
                 })
             },
-            &mut |_err| (),
+            &mut |err| error = Some(err),
         ) {
-            self.def_collector.unexpanded_macros.push(MacroDirective {
-                module_id: self.module_id,
-                ast_id,
-                legacy: Some(macro_call_id),
-                depth: self.macro_depth + 1,
-            });
+            Ok(Ok(macro_call_id)) => {
+                self.def_collector.unexpanded_macros.push(MacroDirective {
+                    module_id: self.module_id,
+                    ast_id,
+                    legacy: Some(macro_call_id),
+                    depth: self.macro_depth + 1,
+                });
 
-            return;
+                return;
+            }
+            Ok(Err(_)) => {
+                // Built-in macro failed eager expansion.
+                self.def_collector.def_map.diagnostics.push(DefDiagnostic::macro_error(
+                    self.module_id,
+                    MacroCallKind::FnLike(ast_id.ast_id),
+                    error.map(|e| e.to_string()).unwrap_or_else(|| String::from("macro error")),
```

Comment:
We have an `ErrorReported` here, so `error` must be `Some`:

```suggestion
                    error.unwrap().to_string(),
```